File: tic-tac-toe/python/tests/board_builder.py
from tic_tac_toe import BOARD_SIZE, Board, Cell


class BoardBuilder:
    def __init__(self) -> None:
        self._grid = [[Cell.EMPTY for _ in range(BOARD_SIZE)] for _ in range(BOARD_SIZE)]

    def with_x_at(self, row: int, col: int) -> "BoardBuilder":
        self._grid[row][col] = Cell.X
        return self

    def with_o_at(self, row: int, col: int) -> "BoardBuilder":
        self._grid[row][col] = Cell.O
        return self

    def build(self) -> Board:
        return Board.from_grid(self._grid)

File: tic-tac-toe/typescript/tests/boardBuilder.ts
import { BOARD_SIZE, Board, type Cell } from '../src/board.js';

export class BoardBuilder {
  private readonly grid: Cell[][] = Array.from({ length: BOARD_SIZE }, () =>
    Array.from({ length: BOARD_SIZE }, () => 'Empty' as Cell),
  );

  withXAt(row: number, col: number): this { this.grid[row]![col] = 'X'; return this; }
  withOAt(row: number, col: number): this { this.grid[row]![col] = 'O'; return this; }

  build(): Board { return Board.fromGrid(this.grid); }
}
